Guard against memory unmapping errors

mhvk opened the pull request:
This is an update of #4585, which ensures we don't get stack traces from badly corrupted compressed data hdus. It makes an additional change so that the code actually works as advertised (initializing `*buf=NULL`), simplifies it a bit, adds an extra error message, and adds tests.

fixes #3118

mhvk commented:
@MSeifert04, @pllim - I think this is OK -- at least it provable avoids getting a complete breakdown, stacktrace included -- but a closer look is most welcome...

@embray - if you have time, could you check that this indeed fixes #3118?

MSeifert04 commented on astropy/io/fits/src/compressionmodule.c:
```diff
@@ -764,6 +768,13 @@ void init_output_buffer(PyObject* hdu, void** buf, size_t* bufsize) {
     *buf = calloc(*bufsize, sizeof(char));
+    if (*buf == NULL) {
+      // Checking if calloc failed.
+      PyErr_SetString(PyExc_TypeError,
```
This should be a `PyExc_MemoryError` and the next line has inconsistent indentation.

mhvk replied:
OK, makes sense, done.

MSeifert04 commented on astropy/io/fits/src/compressionmodule.c:
```diff
@@ -888,7 +899,7 @@ PyObject* compression_compress_hdu(PyObject* self, PyObject* args)
     tcolumn* columns = NULL;
 
-    void* outbuf;
+    void* outbuf = NULL;
```
Is this needed? I'm only asking out of curiosity.

mhvk replied:
Yes, it turned out that was in fact the most important change I still had to make -- the problem is that `init_output_buffer` does not touch `outbuf` at all when an error occurs and then the comparison with `NULL` fails, leading to a later trial to unmap `buf`.

(I guess an alternative would be to check for `PyErr_Occurred` or to let `init_output_buffer` return or set a status flag.)

pllim replied:
I don't see anything wrong just setting it to NULL anyway.

mhvk replied:
Agreed; overall, it seemed best to just keep any individual piece of code as little reliant as possible on the rest being correct...

MSeifert04 commented on CHANGES.rst:
```diff
@@ -204,6 +204,9 @@ Bug Fixes
   - Fix out-of-order TUNITn cards when writing tables to FITS. [#5720]
 
+  - Guard against extremely unlikely problems in compressed images, which
+    could lead to memory unmapping errors. 
```
PR number is missing. :)

mhvk commented:
OK, I fixed the indentation and added the PR number to the changelog. (I used `[skip ci]` since this should not change anything, but checked it locally just to be sure).

Ready to go in?

pllim commented:
LGTM :+1:   Thanks!

MSeifert04 commented:
👍 Thanks
